sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -30,211 +30,382 @@ import org.apache.spark.util.Utils
+
+        case n: NaNvl =>
+          updateWithExpr(n.left, evalCount, condEvalCount)
+          updateWithExpr(n.right, 0, (evalCount + condEvalCount) / 2)
+
+        case c: Coalesce =>
+          updateWithExpr(c.children.head, evalCount, condEvalCount)
+          var cec = evalCount + condEvalCount
+          c.children.tail.foreach {
+            cec /= 2
+            updateWithExpr(_, 0, cec)
+          }
+
+        case e => e.children.foreach(updateWithExpr(_, evalCount, condEvalCount))
+      }
     }
   }
 
+  private def intersectWithExpr(
+      expr: Expression,
+      evalCount: Int,
+      condEvalCount: Double) = {
+    val localMap = new EquivalentExpressions
+    localMap.updateWithExpr(expr, evalCount, condEvalCount)
+    intersectWith(localMap)
+  }
+
   /**
-   * Adds the expression to this data structure recursively. Stops if a matching expression
-   * is found. That is, if `expr` has already been added, its children are not added.
+   * This method can be used to compute the equivalence map if there is a branching in expression
+   * evaluation.
+   * E.g. if we have `If(_, a, b)` expression and `A` and `B` are the equivalence maps built from
+   * `a` and `b` this method computes the equivalence map `C` in which the keys are the superset of
+   * expressions from both `A` and `B`. The `evalCount` statistics of expressions in `C` depends on
+   * whether the expression was present in both `A` and `B` or not.
+   * If an expression was present in both then the result `evalCount` of the expression is the
+   * minimum of `evalCount`s from `A` and `B` (intersection of equivalence maps).
+   * For the sake of simplicity branching is modelled with 0.5 / 0.5 probabilities so the
+   * `condEvalCount` statistics of expressions in `C` are calculated by adjusting both
+   * `condEvalCount` from `A` and `B` by `0.5` and summing them. Also, difference between
+   * `evalCount` of an expression from `A` and `B` becomes part of `condEvalCount` and so adjusted
+   * by `0.5`.
+   *
+   * Please note that this method modifies `map` and `otherMap` is no longer safe to use after this
+   * method.
    */
-  def addExprTree(
-      expr: Expression,
-      map: mutable.HashMap[ExpressionEquals, ExpressionStats] = equivalenceMap): Unit = {
-    if (supportedExpression(expr)) {
-      updateExprTree(expr, map)
+  private def intersectWith(other: EquivalentExpressions) = {
+    inflate()
+    other.inflate()
+
+    val zippedMaps = maps.zip(other.maps)
+    zippedMaps.foreach { case (map, otherMap) =>
+      map.foreach { case (key, value) =>
+        otherMap.remove(key) match {
+          case Some(otherValue) =>
+            val (min, max) = if (value.realEvalCount < otherValue.realEvalCount) {
+              (value.realEvalCount, otherValue.realEvalCount)
+            } else {
+              (otherValue.realEvalCount, value.realEvalCount)
+            }
+            value.realCondEvalCount += otherValue.realCondEvalCount + max - min
```

Comment:
shouldn't this be `value.realCondEvalCount = (value.realCondEvalCount + otherValue.realCondEvalCount) / 2`?

Reply:
`value.realCondEvalCount = (value.realCondEvalCount + otherValue.realCondEvalCount + max - min) / 2` is the full calculation, but the `value.realCondEvalCount /= 2` extracted a bit below.

The `max - min` / 2 is also need. E.g. if we have `If(_, a + b, (a + b) + (a + b))` then during the intersect of the `then` and `else` branches we have `a + b -> 1 + 0` in `then` and `a + b -> 2 + 0` in `else`. The result should be (`a + b -> 1 + 0.5)`.